### ISSUE TITLE
question about mpileup read-pair overlap removal

### ISSUE THREAD

**Issue by solveforj** [samtools/samtools#1985]
The documentation for samtools mpileup states under the `-x, --ignore-overlaps-removal, --disable-overlap-removal` option:

> When enabled, where the ends of a read-pair overlap the overlapping region will have one base selected and the duplicate base nullified by setting its phred score to zero. It will then be discarded by the --min-BQ option unless this is zero.

I would appreciate some clarifications on the terminology here. 

First, does `--min-BQ` filter bases using their BAQ or their original phred base quality score from the FASTQ file, assuming that BAQ is computed? I am assuming that BAQ is what is being used.

Second, does this mean that duplicate bases are nullified by setting their _original base qualities_ (phred scores) in the SAM/BAM file to zero **prior** to BAQ computation? Or, is BAQ **first** computed and then duplicate bases are nullilfied by setting their BAQ scores to zero?

**Comment by jkbonfield** [comment]
Good question regarding BAQ, and not easy to work out from the code either as it's quite a tangle.  However fortunately it's easy to test:

```
$ samtools mpileup -f $HREF -Q30  ~/scratch/data/9827_2#49.1m.bam|awk '{print $NF}'|wc -c
[mpileup] 1 samples in 1 input files
13151475
```

```
$ samtools mpileup -f $HREF -Q30 -B ~/scratch/data/9827_2#49.1m.bam|awk '{print $NF}'|wc -c
[mpileup] 1 samples in 1 input files
13212497
```

So -B changed the number of quality values output when filtering by QUAL.  At `-Q0` we get the same value both times (14457662) so it's clear that BAQ modifies quality before -Q filtering is applied.

For the second question, putting break points on `sam_prob_realn` and `overlap_push` I see it always calls `sam_prob_realn` first for each read.  So I conclude it's doing BAQ before the overlap removal.

None of this is clear though, and the code isn't obvious.

One thing though, IMO I'd just side-step the entire issue by simply disabling BAQ.  It's not proved to be a particularly reliable thing and has been a continual cause of bug reports (why isn't this obvious variant called?).   It's simply too blunt an instrument.  I'd probably have advocated for making off the default were it not for causing breakage of pipelines.  In bcftools we disabled it in most cases now, only enabling it when apparently necessary, because it caused way too many false negatives.  In my latest bcftools work I disabled it completely (in a new `-X` profile) and replaced it with STR analysis which appears to be considerably more robust.
